packages/aragon-cli/src/commands/apm_cmds/util/generate-artifact.js

```diff
@@ -0,0 +1,215 @@
+const fs = require('fs')
+const path = require('path')
+const { readJson, writeJson } = require('fs-extra')
+const extract = require('../../../helpers/solidity-extractor')
+const namehash = require('eth-ens-namehash')
+const { keccak256 } = require('js-sha3')
+
+const { ARTIFACT_FILE } = require('./preprare-files')
+const SOLIDITY_FILE = 'code.sol'
+
+const getMajor = version => version.split('.')[0]
+
+function decorateFunctionsWithAbi(functions, abi, web3) {
+  functions.forEach(f => {
+    f.abi = abi.find(
+      interfaceObject =>
+        web3.eth.abi.encodeFunctionSignature(interfaceObject) ===
+        web3.eth.abi.encodeFunctionSignature(f.sig)
+    )
+  })
+}
+
+async function deprecatedFunctions(apm, artifact, web3, reporter) {
+  let deprecated = {}
+  try {
+    const deprecatedFunctionsSig = new Set()
+    const versions = await apm.getAllVersions(artifact.appName)
```

Comment:
Didn't notice this before, but we probably want to do some error handling and optimization around this.

`apm.getAllVersions()` is quite expensive; it'll fetch the IPFS content for each version. An optimization we may want to make is to not use `apm.js`, but something like https://github.com/aragon/aragon.js/blob/master/packages/aragon-wrapper/src/core/apm/repo.js#L9 to extract all the on-chain information first, and only load the IPFS content we actually need (the last version of each major range).

We should also do a bit of error handling, where if this fails, we prompt the user if they'd like to continue anyway. Not all upgrades will have deprecations, and it'd be nice if the user could choose to proceed to publishing even if they couldn't fetch the last published version's content.

Reply:
Make sense. I'll take the opportunity and also move `generate artifact` task to their own command. I think is better to separate the concerns here and not having it in the `publish` command